### PR TITLE
GUAC-1388 Allow LDAP role-based access control for guacConfigGroups

### PR DESCRIPTION
**Pull request by richarpw**

### CODE REVIEW

**Comment by lorengordon** [comment]
Saw the branch was out of date and rebased. Please let us know if you have any questions. Thanks!

**Comment by lorengordon** [comment]
Rebased again to keep the branch in sync with master. @mike-jumper, anything more you need on this PR?

**Comment by mike-jumper** [comment]
@lorengordon - thanks. I don't have any input at present, but have not had the time to look over this with any depth recently. We are putting this in our next sprint, however, so review and comments from our side will commence very shortly.

**Comment by lorengordon** [comment]
Gotcha, thanks @mike-jumper. I admit to hoping this would make it into 0.9.9, but nbd. We'll just keep using a customized version of the plugin internally for now. Easy enough. Fyi, we've been using this idea with 0.9.7 for the last month or so and it's been solid (and super convenient for enabling role-based access to guac connections). Definitely let us know if you see any issues!

File: extensions/guacamole-auth-ldap/doc/examples/exampleConfigGroup.ldif [hunk 1]
```diff
@@ -9,3 +9,4 @@ guacConfigParameter: port=5900
 guacConfigParameter: password=secret
 member: cn=user1,dc=example,dc=com
 member: cn=user2,dc=example,dc=com
+seeAlso: ou=admins,ou=groups,dc=example,dc=com
```
Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/doc/examples/exampleConfigGroup.ldif]: Is the `seeAlso` attribute a standard attribute inherited from the definition of one of the object classes already used by `guacConfigGroup`?

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java [hunk 2]
```diff
@@ -109,11 +109,38 @@
...
+	                    LDAPConnection.SCOPE_SUB,
+	                    "(&(!(objectClass=guacConfigGroup))(member=" + escapingService.escapeLDAPSearchFilter(userDN) + "))",
+	                    null,
+	                    false
+	            );
+
+	            // Create filter with seeAlso groups, the seeAlso attribute on
+	            // the guacConfigGroup will refer to other groups, any members
+	            // in these groups have access to the connection also
+	        	while (userRoleGroupResults.hasMore()) {
```
Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: Here and elsewhere, please do not use tabs. All code should use 4-space indent.

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java [hunk 3]
```diff
@@ -109,11 +109,38 @@
             // possibly be null
             assert(userDN != null);
 
-            // Find all Guacamole connections for the given user
+            // If group base DN is specified search for user groups
+            String groupBaseDN = confService.getGroupBaseDN();
+            // Create a separate filter for user groups
```
Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: While you're fixing the indentation, mind maybe adding a blank line above here? This looks rather jumbled as-is. Everything is just squished together.

Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: To clarify a bit, it's odd to see something like:

```java
// If group base DN is specified search for user groups
String groupBaseDN = confService.getGroupBaseDN();
// Create a separate filter for user groups
StringBuilder userRoleGroupFilter = new StringBuilder();
if (groupBaseDN != null) {
```

Normally, you would expect a blank line to separate two logical paragraphs of code like this.

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java [hunk 4]
```diff
@@ -109,11 +109,38 @@
...
+
+        	// Find all Guacamole connections for the given user by
+        	// looking for direct membership in the guacConfigGroup
+        	// and possibly any groups the user is a member of that are
+            // referred to in the seeAlso attribute of the guacConfigGroup
             LDAPSearchResults results = ldapConnection.search(
                 configurationBaseDN,
                 LDAPConnection.SCOPE_SUB,
-                "(&(objectClass=guacConfigGroup)(member=" + escapingService.escapeLDAPSearchFilter(userDN) + "))",
+                "(&(objectClass=guacConfigGroup)(|(member=" + escapingService.escapeLDAPSearchFilter(userDN) + ")" + userRoleGroupFilter.toString() + "))",
```
Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: It's very odd to so nicely avoid string concatenation and use a `StringBuilder` to produce the group filter ... only to then use that `StringBuilder` within string concatenation.

Perhaps there is a more elegant way to produce this? Maybe using only one `StringBuilder` for the filter overall?

**Comment by mike-jumper** [comment]
@lorengordon - results of first code review are above.

Looking at our records, we have a CLA on file for @richarpw, but not for yourself. The CLA covers a particular individual's contributions, so if you are both working for the same company (as looks to be the case), you would need to either (1) sign a CLA to cover your own contributions going forward (which would include whatever you have contributed on top of what @richarpw has already done) or (2) have someone with sufficient authority at your company sign a CLA to cover contributions on behalf of that company.

The CLA is here: http://glyptodon.org/cla.html

**Comment by lorengordon** [comment]
@mike-jumper, thanks for the review. I'm just tracking the issue because I deployed our implementation of Guacamole and needed this feature in the plugin. My only involvement on the code has been rebasing to keep the branch up-to-date. I'll defer to @richarpw for any of the code. Do you need a CLA from me even if I'm not actually contributing code?

**Comment by mike-jumper** [comment]
@lorengordon - ah, OK. No, if you're not contributing any code, then no CLA is necessary, but then please either do not rebase or let @richarpw do the rebase. Otherwise, the source history will not jive with our own contributor records.

Currently, the git history shows both yourself and @richarpw as the contributors of the change, which could prove legally confusing.

File: extensions/guacamole-auth-ldap/doc/examples/exampleConfigGroup.ldif (hunk 1 above)
Comment (richarpw) [inline extensions/guacamole-auth-ldap/doc/examples/exampleConfigGroup.ldif]: Yes it comes from groupOfNames: http://www.zytrax.com/books/ldap/ape/#groupofnames

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java (hunk 3 above)
Comment (richarpw) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: Will add a blank line.

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java (hunk 2 above)
Comment (richarpw) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: Will fix the tabs, thought they were gone, will re-check the files.

File: extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java (hunk 4 above)
Comment (richarpw) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: Frankly I think it is easier to read like this but I will use a StringBuilder for the entire filter.

Comment (mike-jumper) [inline extensions/guacamole-auth-ldap/src/main/java/org/glyptodon/guacamole/auth/ldap/connection/ConnectionService.java]: Readability is definitely extremely important. Perhaps there is a way that is clean *and* doesn't concatenate a `StringBuilder`?

Off the top of my head, moving the filter generation to its own private function might help smooth things, but I'm sure there are any number of ways to achieve this.

**Comment by richarpw** [comment]
@mike-jumper, take a look at the updated code when you get a chance and let me know if you need any other changes

**Comment by mike-jumper** [comment]
Thanks, @richarpw. This looks excellent. I have some minor additional changes that I think should be made, but they're not worth additional back-and-forth. I will get these changes moved forward via #323.

**Pull request closed without merging**